tests/test_simpleButler.py

```diff
@@ -882,10 +883,61 @@ def makeButler(self, writeable: bool = False) -> Butler:
         registryConfig = RegistryConfig(config.get("registry"))
         _RegistryFactory(registryConfig).create_from_config()
 
+        # Write the YAML file so that some tests can recreate butler from it.
+        config.dumpToUri(os.path.join(self.root, "butler.yaml"))
         butler = Butler.from_config(config, writeable=writeable)
         DatastoreMock.apply(butler)
         return butler
 
+    def test_dataset_uris(self):
+        """Test that dataset URIs can be parsed and retrieved."""
+        butler = self.makeButler(writeable=True)
+        butler.import_(filename=os.path.join(TESTDIR, "data", "registry", "base.yaml"))
+        butler.import_(filename=os.path.join(TESTDIR, "data", "registry", self.datasetsImportFile))
+
+        butler.registry.defaults = RegistryDefaults(collections=["imported_g"])
+        ref = butler.find_dataset("flat", detector=2, physical_filter="Cam1-G")
+        self.assertIsInstance(ref, DatasetRef)
+
+        # Get the butler root for the URI.
+        config_dir = butler._config["root"]
+
+        # Read it via a repo label and a path.
+        with tempfile.NamedTemporaryFile(mode="w", suffix=".yaml") as index_file:
+            label = "test_repo"
+            index_file.write(f"{label}: {config_dir}\n")
+            index_file.flush()
+            with mock_env({"DAF_BUTLER_REPOSITORY_INDEX": index_file.name}):
+                butler_factory = LabeledButlerFactory()
+                factory = butler_factory.bind(access_token=None)
```

Comment:
@rra does this approach work for you in the cutout service?

```python
factory = butler_factory.bin(access_token=token)
...
ref = Butler.get_dataset_from_uri(dataset_uri, factory=factory)
```
?

Reply:
The cutout service passes a Butler instance into the backend, so the pattern looks like:
```python
butler_factory = LabeledButlerFactory()

def _get_backend(label: str, token: str) -> ImageCutoutBackend:
    # Called for each cutout
    factory = butler_factory.bind(access_token=token)
    butler = factory.create_butler(label=label)
    # ...
    return ImageCutoutBackend(butler, projection_finder, output, tmpdir)
```
Is that what you had in mind? Basically move the access token parameter to `create_butler` to an intermediate step to create a factory with a bound token? If so, that would be fine here.

Reply:
Ah. No. It's probably the wrong API for you. Somewhere you are parsing the `butler://` URI and I'm trying to provide code here that will hide the URI structure from you (so that we can also support the new `ivo://` URIs). This PR creates two APIs: one that parses the URI and returns the butler repo label and the UUID, and another API (that is the one I talk about here) that lets you retrieve the DatasetRef directly from the URI and a butler factory. Maybe `get_dataset_from_uri` should return the `Butler` instance along with the `DatasetRef`?

Reply:
Maybe I could pass the return value of `bind` into the backend along with the URIs as-is without parsing them and then the backend can do whatever it needs to do? That would be even more convenient. In other words, have the constructor of `ImageCutoutBackend` take a Butler factor instead of a Butler instance.

Reply:
I think passing the bound factory and URI into whatever is wanting to know the DatasetRef is what we want here.